### DISCUSSION TITLE
Berry Script: Halleffect/ADC/touch - Faster Switching as Rules?

### DISCUSSION THREAD

**Discussion by ozett** [arendst/Tasmota#18116] [Q&A]
in berry you can read sensors faster than 1 sec. interval
i tried to use the built in halleffect (without berry) only with rules to trigger a relay and it had a longer delay.
```
----rules
#"ESP32":{"Temperature":53.3,"HallEffect":63},"TempUnit":"C"}

Rule1 ON ESP32#HallEffect < 0  DO Power1 1 ENDON
Rule2 ON ESP32#HallEffect > 0  DO Power1 0 ENDON

rule1 1  # activate
rule2 1  # activare

```
maybe reading time of the internal hall-sensor is limited throug implementation?
https://github.com/arendst/Tasmota/blob/c1ea8953cb3f496a5dbf20033a665ad3ad26f136/tasmota/tasmota_xsns_sensor/xsns_127_esp32_sensors.ino#L38

maybe it wasn't the right approach with rules and rules have a builtin minimal delay-time?

## Berry
and after some ideas from other threads 

[hall in berry](https://github.com/arendst/Tasmota/discussions/13741)
[ADC in berry](https://github.com/arendst/Tasmota/discussions/18046)
[tasmota.read_sensors()](https://github.com/arendst/Tasmota/pull/13756)
[diagram of sensor reading](https://esp32.com/viewtopic.php?f=19&t=4963)

i put something together in berry what works quite well now, ([and uses timer-callback](https://github.com/arendst/Tasmota/discussions/18046#discussioncomment-5203427))
 the relay switches fairly quickly (compared to the rules trigger) when the magnet now comes closer to the ESP core.
even with this (longer) callbacktime of 1000ms (!)

```
# HALL EFFECT, quicker reaction in berry than with rules
---
import json

def func2()
  # print('func2: ', tasmota.millis())
  sensors=json.load(tasmota.read_sensors())
  if sensors['ESP32']['HallEffect'] < 0
    print('kleiner null: ', sensors['ESP32']['HallEffect'])
    
    gpio.digital_write(gpio.pin(gpio.REL1), gpio.HIGH)  
  else
    print(sensors['ESP32']['HallEffect'])
    gpio.digital_write(gpio.pin(gpio.REL1), gpio.LOW)  
  end


end

def set_timer_modulo(delay,f,id)
  # nothing elst to do here
  # set timer, building a loop 
  tasmota.set_timer(delay, def()set_timer_modulo(delay,f,id); func2() end, id)
end

set_timer_modulo(1000, def () end, 'Ticker')
# tasmota.remove_timer('Ticker')
```
maybe a pro can improve the skeleton, i'm a total beginner. 
or there [will come up ](https://github.com/arendst/Tasmota/discussions/18046#discussioncomment-5203416)a berry tasmota.function() like `gpio.analog_read and gpio.hall_read` to make use of?

I would be happy about any suggestions or improvements.

**Comment by ozett** [comment]
tested the above code with the hallsensor to my satisfaction ,
i feel encouraged to do the same thing with the touch-sensor.
https://tasmota.github.io/docs/TouchPin/

but 😭 not possible? 
is there access to the `button_tc` values like to the hallsensor from berry? berry.gpio?

Surprise: having the hallsensor-rules still in place from my experiments, i trigger the relay while touching the wire connected to the button_tc GPIO4 (touch1).   is this a bug? or a magnetic side effect? i will investigate this with reboot. and after the reboot the relay is still toggled from the touch-pin. WHATS THIS? normal?

edit: [seems normal
](https://tasmota.github.io/docs/Buttons-and-Switches/#switch) What a surprise... <kbd>![image](https://user-images.githubusercontent.com/5937811/222960460-ea31e3bd-3225-476d-a6a7-fd7310bce968.png)

**Reply by sfromis** [reply]
Yeah, a Button is a Button, also when touch. And buttons results in default actions.
Maybe also calibrate.... https://tasmota.github.io/docs/TouchPin/

**Reply by ozett** [reply]
works like magic, great, no need for berry-fiddling for the touch-buttons. Alreday done by Magicians.
leaves me room to concentrate on the hall-reading-routine and waiting for feedback to improve my berry-hacking

**Reply by ozett** [reply]
maybe another idea ?
what if tasmota could define a hallsensor level (below/above a certain number) as a button?
that would easily create a magnetic triggert button with default actions...

**Comment by sfromis** [comment]
When it comes to your code, it seems that you really aim for making it confusing. You supply a dummy function to `def set_timer_modulo` and modify it to not use the supplied function, but instead hardcode a call of the function you should have provided.

It would make more sense with `def set_timer_modulo` from the docs and `set_timer_modulo(1000, func2, 'Ticker')`. That's how it is meant to be used. Important, do not use `func2()` there, as that would execute the function, instead of passing it.

My preference would also be something like `var magvalue = sensors['ESP32']['HallEffect']` once only, and then using the var instead of repeatedly referring to the map structure. Not terribly important, but does make the code a bit smaller and more understandable.

And instead of `gpio.digital_write(gpio.pin(gpio.REL1), gpio.HIGH)`, I'd prefer using the specific Berry feature for switching a relay output, `tasmota.set_power(0, true)`. Again shorter, more readable, and better integrated with how Tasmota works.

Except for the print() you're using, you could also skip the whole if, and do something like:
`tasmota.set_power(0, magvalue<0)`
The trick is that set_power tales a boolean value, which can be the result of doing a comparison.

Basically, your func2 boils down to.
```
def func2()
  import json
  tasmota.set_power(0, json.load(tasmota.read_sensors())['ESP32']['HallEffect'] < 0)
end
```
I prefer doing `import json` where relevant, not at the global scope. Matter of taste, and more robust when changing stuff.

**Reply by ozett** [reply]
two realizations: 
i will **_never_** learn to code `and` this small code works great to perform as a button.

edited
```
def func2()
  import json
  tasmota.set_power(0, json.load(tasmota.read_sensors())['ESP32']['HallEffect'] < 0)
end

def set_timer_modulo(delay,f,id)
  # set timer, building a loop 
  tasmota.set_timer(delay, def()set_timer_modulo(delay,f,id) f() end, id)
end

set_timer_modulo(1000, func2, 'Ticker')
# tasmota.remove_timer('Ticker')
```

is this better styled? after seeing the default magic of buttons the magnetic hall-sensor should have a threshold to become a button, too.
and, by the way: isn't the `import json` not unneccessary repeated every time the function is called?

**Reply by sfromis** [reply]
`import json` does not apply at run time, it is a compile-time directive telling how to resolve names. 

BTW, you can make it more compact still. Not strictly recommending this, more of an illustration.
```
import json
set_timer_modulo(1000, /->tasmota.set_power(0, json.load(tasmota.read_sensors())['ESP32']['HallEffect'] < 0), 'Ticker')
```
That's all, folks! :duck:

**Reply by ozett** [reply]
Magic Code: 
having your improved elegant little script running,
the touch-button function dont toggle anymore. 

everytime i touch the button to ON, 
it switches automagically and instantly to OFF

i proved it with rebooting and without berry-code: touch-button is normal.
after activating this berry-code in the console, the touch-button toggle does not hold its state. 
touching brings the relay to ON and instantly after that automaticall to OFF.

i will leave this weired stuff as it is, enough great learning for me with berry on this sunday..
thanks a lot
